### ISSUE TITLE
[Bug] MiniMax-Music3: silent dropout at final-chunk stitching boundary

### ISSUE THREAD

**Issue by Plaidstallion** [sgl-project/sglang-omni#1542]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang-omni/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

Generating audio with MiniMax-Music3 via `sgl-omni serve`, I'm seeing a reproducible near-total silent gap in the output audio, occurring specifically at the handoff into the final acoustic chunk.

Measured directly on the output WAV (RMS amplitude in 0.25s windows): amplitude drops to noise-floor level (~74–79 out of 0–32767) from 11.25s to 13.75s, a clean 2.5-second silent gap, then jumps immediately back to full amplitude (~7400+) at 14.00s. Real audio exists both before and after the gap.

Server log shows the dropout timing correlates directly with the chunk-stitching boundary: chunk 1 covers 4.0–12.0s and the final chunk (idx=2, `final=True`) covers 8.0–16.0s, with their overlap region at 8.0–12.0s. The dropout sits right at the tail of that overlap and bleeds into the final chunk's exclusive range, pointing at the crossfade/stitching logic specifically when handling the transition into the final chunk.

### Reproduction

Request:

`max_new_tokens: 400` (16s of audio)
Prompt: simple, unstructured (`"Generate an 8 bar loop. make it funky"`)
`finish_reason=length` (completed normally, not an early stop)

Server log, acoustic chunks:

```
idx=0 frames=200 frame_range=[0,200) final=False
idx=1 frames=200 frame_range=[100,300) final=False
idx=2 frames=200 frame_range=[200,400) final=True
```

[minimax-music3-chunk-stitch-dropout.wav](https://github.com/user-attachments/files/31088066/minimax-music3-chunk-stitch-dropout.wav)
[minimax-music3-chunk-stitch-dropout-log.txt](https://github.com/user-attachments/files/31088065/minimax-music3-chunk-stitch-dropout-log.txt)

### Environment

Single GPU: RTX 3090 (24GB)
`--mem-fraction-static 0.85`
Otherwise default serving config, `sgl-omni serve --model-path MiniMaxAI/MiniMax-Music3`

[minimax-music3-startup-log.txt](https://github.com/user-attachments/files/31088315/minimax-music3-startup-log.txt)

**Comment by AkazaAkane** [comment]
Reproduced on current main.

Using an H200, MiniMax-Music3, max_new_tokens=400, and seed 9:
  - Chunks: [0,200), [100,300), [200,400) final=True
  - finish_reason=length
  - Output duration: 16.01 seconds
  - RMS drops to roughly 75–79 from 10.00–13.00 seconds
  - RMS immediately recovers to 3603 at 13.00 seconds

Will investigate that.

**Comment by AkazaAkane** [comment]
## Problem

MiniMax-Music3 generated a reproducible near-total silent interval around global **10–13 s** for a 400-frame / 16-second request.

The original issue suspected a bug around the transition into the final acoustic chunk:

```text
chunk 0: [0,200)
chunk 1: [100,300)
chunk 2: [200,400) final=True
```

### Initial Observation
* **~10–13 s:** RMS ~75–80
* **After ~13 s:** RMS immediately recovers to thousands

The investigation has now ruled out stitching/final-chunk handling and localized the silence to a specific region of the AR-produced hidden trajectory.

---

## Investigation & Test Matrix

### 1. Initial Reproduction
Reproduced on current `main` on H200 with:
* **Model:** MiniMax-Music3
* **Seed:** 9
* **max_new_tokens:** 400
* **finish_reason:** length

**Observed:**
* **Chunks:**
  * `[0, 200)`
  * `[100, 300)`
  * `[200, 400)` (`final=True`)
* **Output duration:** ~16 s
* **RMS near noise floor:** around global 10–13 s

This matched the reported RTX 3090 failure closely enough to confirm the issue was reproducible across hardware.

---

### 2. Tested: Latent Continuation / Prompt Boundary

#### Hypothesis
The silence might be introduced when the final acoustic chunk transitions from the carried latent prompt into newly generated DIT latent.

#### Tests
Decoded chunk 2 with:
* **A. Normal continuation:**
  * `initial_latent = previous latent`
  * `initial_condition = previous condition`
* **B. No continuation:**
  * `initial_latent = None`
  * `initial_condition = None`
* **C. Latent only:**
  * `initial_latent = previous latent`
  * `initial_condition = None`

Also changed carried prompt length:
* 86 mel frames
* 172 mel frames *(current behavior)*
* 258 mel frames

#### Result
All variants retained the same raw chunk-2 silence:
* Local ~2–5 s
* Changing prompt length did not move the dropout.

> **Conclusion: Ruled out.**  
> The silence is not caused by the latent-prompt → free-DIT transition.

---

### 3. Tested: Terminal Final-Window Flush

#### Hypothesis
Chunk 2 might be corrupted because `[200, 400)` is emitted through the terminal flush path when generation stops exactly at 400 frames.

#### Hidden Overlap Test
Compared the shared hidden interval:
```python
torch.equal(
    chunk1[:, 100:200],
    chunk2[:, 0:100],
)
```
**Result:**
* `True`
* `max_abs_diff = 0`
* `mean_abs_diff = 0`

#### 400 vs 401 Frame Test
* **At 400 frames:** chunk 2 `[200, 400)` is terminal-flushed.
* **At 401 frames:** chunk 2 `[200, 400)` is emitted through the normal ready-window path.

**SHA-256 Hashes:**
* **400-frame chunk2:** `678dea7b903df2bc8d27d6a2433527de464b52afa8a443e57c4ac7f5add8eb0e`
* **401-frame chunk2:** `678dea7b903df2bc8d27d6a2433527de464b52afa8a443e57c4ac7f5add8eb0e`

*(Outputs are identical)*

> **Conclusion: Ruled out.**  
> The terminal flush path does not corrupt the final hidden chunk.

---

### 4. Tested: DIT Per-Chunk RNG

#### Hypothesis
The deterministic noise seed associated with `chunk_idx=2` might produce a pathological DIT result.

#### Test
Decoded the same chunk-2 hidden input using:
* `chunk_idx = 0`
* `chunk_idx = 1`
* `chunk_idx = 2`
* `chunk_idx = 3`

#### Result
* All four versions retained the same twelve sub-100-RMS windows over the silent region.
* The surrounding waveform amplitudes changed, but the silence itself stayed in place.

> **Conclusion: Ruled out.**  
> The dropout is not specific to the `chunk_idx=2` random noise realization.

---

### 5. Tested: Crop / Concatenation / Resampling

#### Hypothesis
The raw acoustic chunks might be healthy and silence could be introduced later during cropping, concatenation, or 44.1 → 32 kHz resampling.

#### Measurements (Raw Chunk 2)
* **Local 1.75–2.00 s:** RMS 177
* **Local 2.00–2.25 s:** RMS 75
* **Local 2.25–5.00 s:** RMS ~78–80
* **Local 5.00–5.25 s:** RMS 3698

Compared:
1. Raw DAV chunk output
2. Assembled 44.1 kHz waveform
3. Final 32 kHz waveform

The same silent interval exists at every stage.

> **Conclusion: Ruled out.**  
> Cropping, `torch.cat`, and resampling do not introduce the silence.

---

### 6. Tested: Compilation / Numerical Execution Mode

#### Hypothesis
`torch.compile`, TF32, or reduced matmul precision might cause the DIT collapse.

#### Tests
Ran with:
* `compile_acoustic = False`
* TF32 disabled
* `float32` matmul precision = `highest`

#### Result
The same silent interval remained with essentially identical rounded RMS behavior.

> **Conclusion: Ruled out for this symptom.**  
> The dropout is not caused by the compiled acoustic path or TF32 precision.

---

### 7. Key Overlap Observation

The decisive observation was that the silence appears in **both overlapping acoustic windows at the same global time**:
* **Chunk 1 raw local 6–8 s** $\rightarrow$ silence
* **Chunk 2 raw local 2–4 s** $\rightarrow$ silence
* Both correspond to approximately: **global 10–12 s**

Chunk 1 already predicts the future silence in its normally discarded overlap tail, before chunk 2 is decoded.

> **Diagnosis Update:**  
> Shifted from **"final-chunk stitching bug"** $\rightarrow$ **"conditioning-aligned acoustic silence"**.

---

## 8. Causal Hidden-State Experiments

Tested whether a specific AR hidden interval causes the acoustic silence.
* **Responsible region:** approximately global `H[250:325]`
* At 25 AR frames/s, this corresponds to **10–13 s**.

#### Replacement Test
* **Original chunk:** `[200, 400)`
* **Replacement:** Replace `H[250:300]` with known-audible `H[150:200]`.
* **Result:**
  * Local 2–4 s becomes audible.
  * Local 4–5 s remains silent (corresponding to untouched `H[300:325]`).

#### Transplant Test
* Moved problematic `H[250:300]` into chunk 0 at local frames `50–100`.
* **Result:** A new local ~2–4 s silent interval appears. The silence directly follows the hidden-state segment.

#### Shifted-Window Test

| Hidden Window | Local Silence | Global Silence |
| :--- | :--- | :--- |
| `[200, 400)` | 2–5 s | 10–13 s |
| `[175, 375)` | 3–6 s | 10–13 s |
| `[150, 350)` | 4–7 s | 10–13 s |

> **Conclusion: Causality Established**  
> * The AR hidden interval approximately `H[250:325]` drives the silence.
> * The silence is not attached to acoustic chunk position, stitching position, or final-chunk lifecycle.

---

### 9. Official Acoustic Implementation Parity

The dumped SGLang hidden tensor was passed through the official Diffusers acoustic implementation pinned by the MiniMax-Music3 model card.

#### Condition Encoder
* **Aligned-condition mean absolute difference:** `0.1109`
* **Mean per-timestep cosine similarity:** `0.9518`
* *(Note: Replacing only the condition weights in the legacy SGLang DIT did not remove the silence.)*

#### Full Official Acoustic Stack
Using the fully matched official stack (`condition encoder` + `transformer` + `vocoder`) on the exact dumped chunk-2 hidden tensor reproduced the same **local 2–5 s silence**.

> **Conclusion:**  
> The silence is not specific to SGLang's acoustic implementation.

---

### 10. DIT Latent Parity

Official and SGLang DIT were run with an identical injected initial noise tensor $x_0$.
* Official and SGLang latents diverge numerically from step 1.
* However, **both latents encode the same silent interval**.

> **Conclusion:**  
> The numerical DIT mismatch is real, but it does not explain issue #1542.

---

### 11. DAV / Vocoder Cross-Decoding

Cross-decoded official and SGLang latents with both vocoder implementations:
* `official latent` $\rightarrow$ `official vocoder` vs. `SGLang DAV`: **max sample difference = 0**
* `SGLang latent` $\rightarrow$ `official vocoder` vs. `SGLang DAV`: **max sample difference = 0**

> **Conclusion:**  
> The SGLang DAV implementation is bit-identical to the official vocoder for the tested latents. DAV is fully exonerated.

**Comment by PoTaTo-Mika** [comment]
I'll take a further investigation into it, thanks for @AkazaAkane 's help.

**Comment by PoTaTo-Mika** [comment]
Would you mind showing the prompt you use? This will help me locating the issue :)
